onnxruntime/contrib_ops/cpu/bert/attention.cc

```diff
@@ -219,44 +350,45 @@ Status Attention<T>::PrePack(const Tensor& weights, int input_idx, AllocatorPtr
     return Status::OK();
   }
 
+  const auto* weights_data = weights.Data<T>();
   const size_t input_hidden_size = static_cast<size_t>(weights_dims[0]);
   const size_t hidden_size_x3 = static_cast<size_t>(weights_dims[1]);
   const size_t hidden_size = hidden_size_x3 / 3;
-  const size_t head_size = hidden_size / num_heads_;
-
-  // Bail out if the weights shape has an expected shape.
-  if ((hidden_size == 0) || ((hidden_size % num_heads_) != 0) || (hidden_size_x3 != 3 * hidden_size)) {
-    return Status::OK();
-  }
+  size_t q_hidden_size, k_hidden_size, v_hidden_size;
 
-  const auto* weights_data = weights.Data<T>();
+  if (qkv_hidden_sizes_.size() != 0) {
+    q_hidden_size = qkv_hidden_sizes_[0];
+    k_hidden_size = qkv_hidden_sizes_[1];
+    v_hidden_size = qkv_hidden_sizes_[2];
 
-  packed_weights_size_ = MlasGemmPackBSize(head_size, input_hidden_size);
-  if (packed_weights_size_ == 0) {
-    return Status::OK();
-  }
-
-  const size_t loop_len = static_cast<size_t>(3) * num_heads_;
-  size_t packed_weights_data_size = packed_weights_size_ * loop_len;  // The same size would be computed by AllocArray() below
-  auto* packed_weights_data = static_cast<uint8_t*>(alloc->AllocArray(packed_weights_size_, loop_len));
+    if (q_hidden_size == 0 || k_hidden_size == 0 || v_hidden_size == 0) {
+      return Status::OK();
+    }
 
-  // Initialize memory to 0 as there could be some padding associated with pre-packed
-  // buffer memory and we don not want it uninitialized and generate different hashes
-  // if and when we try to cache this pre-packed buffer for sharing between sessions.
-  memset(packed_weights_data, 0, packed_weights_data_size);
+    if (q_hidden_size % num_heads_ != 0 || k_hidden_size % num_heads_ != 0 || v_hidden_size % num_heads_ != 0) {
```

Comment:
This check is not needed since similar check has been done in CheckInputs.

Reply:
PrePack() is called before Compute() which has CheckInputs() to do this again.

However, we have a case when we don't prepack the weights and still compute the matmul in compute(), which suggests this check is needed in both places unless I am missing something.